compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -784,18 +784,19 @@ public boolean isArrayTypesAssignable(BType source, BType target, Set<TypePair>
         } else if (source.tag == TypeTags.ARRAY) {
```

Comment:
I think we can refactor this method a bit.

In L779 and L782 IMO we can call `isAssignable` instead of `isArrayTypesAssignable`.

AFAICT we can then change the first parameter of this method to a `BArrayType` (we'll probably have to deprecate this and introduce a new method). This'll reduce the number of specific checks and reuse the `isAssignable` logic, which has been tested quite a lot, more.

Reply:
Sorry, I think I don't get this clearly?

The `isArrayTypesAssignable` is getting called by `isAssignable` logic anyway [link](https://github.com/ballerina-platform/ballerina-lang/blob/master/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java#L618).
Is there a point going for all the checks in `isAssignable` when we know eventually it will call `isArrayTypesAssignable` function anyway?

And changing the first parameter to BArrayType makes this specific right?

Reply:
> Is there a point going for all the checks in `isAssignable` when we know eventually it will call `isArrayTypesAssignable` function anyway?
> 

L779 and L782 check assignability for the element types of the first level of arrays. And the element types may or may not be array types, so we can't say for sure it'll be calling `isArrayTypesAssignable`.

Consider 

```ballerina
string[] s = [];
(string|int)[] t = s;
```

For `(string|int)[] t = s;`, `isArrayTypesAssignable` gets invoked first, which then at L779 calls `isArrayTypesAssignable` again with source `string` and target `(string|int)` which eventually ends up calling `isAssignable` like in L807. Instead, we can directly call  `isAssignable` to check assignability of element types.



> And changing the first parameter to BArrayType makes this specific right?

IMO, the name `isArrayTypesAssignable` itself has already made it specific. :) This IS a specific check right?

Basically we call this method only when the source type is an array, to check if it is assignable to a specific type. 

So rhsType is always an array.

- if lhsType is also an array - return `isAssignable(rhsElemType, lhsElemType)`
- if lhsType is a union - return true if for at least one member `isArrayTypesAssignable` is true
- if lhsType is `json` - return `isAssignable(rhsElemType, lhsType)` 
- if lhsType is `anydata` - return `rhsType.isAnydata()` 
- if lhsType is `any` - return `rhsElemType.tag != TypeTags.ERROR`

Reply:
The last check is incorrect, sorry, my bad.

See https://github.com/ballerina-platform/ballerina-lang/pull/22494#discussion_r412675176